lib/engine/game/g_18_india/step/corporate_buy_shares_company.rb

```diff
@@ -0,0 +1,183 @@
+          end
+
+          # TODO: Convert Royal Bond in Phase IV
+          def actions(entity)
+            return [] if entity.nil? || !entity&.corporation?
+            return [] unless entity == current_entity
+
+            # May buy any share from Market (it may may itself from Market) [working]
+            # May buy Private Company from Market / Bank [working]
+            # May buy Railroad Bond from Market / Bank (doesn't count against Cert Limit of 3) [working]
+            # may buy One Proxy Cert from the IPO Row (but NOT itself) [working]
+            # convert a Royal Bond to a share of GIPR (Phase IV && must have space) [TODO]
+            actions = []
+            actions << 'corporate_buy_company' if can_buy_any_companies?(entity)
+            actions << 'corporate_buy_shares' if can_buy_any?(entity)
+            actions << 'pass' if actions.any?
+
+            actions
+          end
+
+          # ----- methods for buying companies (new) -----
+
+          def can_buy_any_companies?(entity)
+            return false if bought?(entity)
+
+            buyable_companies(entity).count.positive?
+          end
+
+          def buyable_companies(entity)
+            return [] unless entity.corporation?
+
+            companies = []
+            companies += @game.bank_owned_companies.select { |c| can_buy_comp_from_market?(entity, c) }
+            companies += @game.top_of_ipo_rows.select { |c| can_buy_comp_from_ipo?(entity, c) }
+
+            companies
+          end
+
+          def can_buy_comp_from_market?(entity, company)
+            return false if bought?(entity)
+            return false if (@game.num_certs(entity) >= @game.cert_limit(entity)) && (company.type != :bond)
+
+            entity.cash >= company.value
+          end
+
+          def can_buy_comp_from_ipo?(entity, company)
+            return false if bought?(entity)
+            return false if @game.num_certs(entity) >= @game.cert_limit(entity)
+            return false if company.type == :share && (entity == company.treasury.corporation) # may NOT buy self from IPO
+
+            _row, index = @game.ipo_row_and_index(company)
+            index.zero? && (entity.cash >= company.value)
+          end
+
+          def can_buy_company?(entity, company)
+            return false if entity.nil?
+            return false if bought?(entity)
+            return false if @game.bank_owned_companies.include?(company) && !can_buy_comp_from_market?(entity, company)
+            return false if @game.in_ipo?(company) && !can_buy_comp_from_ipo?(entity, company)
+
+            entity.cash >= company.value
+          end
+
+          def process_corporate_buy_company(action)
+            entity = action.entity
+            company = action.company
+            price = action.price
+            owner = company.owner
+            what = company.type == :share ? "a share of #{company.name}" : "Private #{company.name}"
```

Comment:
Could you either give this a more descriptive name, or move it down to right by the log statement where it is used, so it's easier to tell what it's for?

Reply:
Renamed and moved just before the log statement.